compiler/rustc_typeck/src/check/method/mod.rs

```diff
@@ -463,6 +538,45 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             expr_id,
             ProbeScope::TraitsInScope,
         )?;
+
+        if span.edition() < Edition::Edition2021 {
+            if let sym::try_into | sym::try_from | sym::from_iter = method_name.name {
+                if let probe::PickKind::TraitPick = pick.kind {
+                    if !matches!(tcx.crate_name(pick.item.def_id.krate), sym::std | sym::core) {
+                        tcx.struct_span_lint_hir(FUTURE_PRELUDE_COLLISION, expr_id, span, |lint| {
+                            let trait_def_id = pick.item.container.assert_trait();
+                            let trait_generics = tcx.generics_of(trait_def_id);
+                            let parameter_count = trait_generics.count() - (trait_generics.has_self as usize);
+
+                            let trait_name = if parameter_count == 0 {
+                                tcx.def_path_str(trait_def_id)
+                            } else {
+                                format!(
+                                    "{}<{}>",
+                                    tcx.def_path_str(trait_def_id),
+                                    std::iter::repeat("_").take(parameter_count).collect::<Vec<_>>().join(", ")
+                                )
+                            };
+
+                            let mut lint = lint.build(&format!(
+                                "trait-associated function `{}` will become ambiguous in Rust 2021",
+                                method_name.name
+                            ));
+
+                            lint.span_suggestion(
+                                span,
+                                "disambiguate the associated function",
+                                format!("<{} as {}>::{}", self_ty, trait_name, method_name.name,),
```

Comment:
Better than using `self_ty` here would be if we can get the span of the type as the user wrote it and just copy-and-paste it. I wonder how hard that is. Have to check the caller!